### ISSUE TITLE
App.route() not matching all variable wildcards

### ISSUE THREAD

**Issue by DariusMiu** [bocadilloproject/bocadillo#272]
**Expected behavior**
variable wildcards match expected queries (`@api.route('{path}')` should match all urls, but does not)

**Actual behavior**
variable wildcards only match some expected queries (the above can match any url with a single slash after the domain, ie `http://example.com/page` or `http://example.com/image.png`)

**To Reproduce**
```python
import bocadillo
api = bocadillo.App()

@api.route('{path}')
async def slash(req, res, path) :
	print('success')

if __name__ == '__main__' :
	from uvicorn.main import run
	run(api, host='127.0.0.1', port=80)
```
FAILS

```python
import bocadillo
api = bocadillo.App()

@api.route('{}')
async def slash(req, res) :
	print('success')

if __name__ == '__main__' :
	from uvicorn.main import run
	run(api, host='127.0.0.1', port=80)
```
SUCCEEDS

**Screenshots/Traceback**
![image](https://user-images.githubusercontent.com/29378233/57196219-b0608000-6f28-11e9-9664-551a4c2cd508.png)

![image](https://user-images.githubusercontent.com/29378233/57196207-99ba2900-6f28-11e9-871b-6436ecce70b5.png)

**Possible solutions**
This might be the expected or intended behavior, but this wasn't how app.route behaved previously, as this is only occurring after updating bocadillo. The error seems like it should be in how bocadillo matches urls to the specified routes.

**Specs**
- OS + Version: OpenSUSE Leap 15.0
- Python Version: 3.6.5
- Bocadillo Version: 0.15.0

**Additional context**
I first discovered this by finding out that my route for images only worked on images in the root folder: `@api.route('/{path}.png')` only caught urls on the root. ie `http://example.com/image.png` versus `http://example.com/images/image.png`.

**Comment by florimondmanca** [comment]
Hey, thanks for reporting this.

I do think this has something to do with a route parameter not able to contain slashes itself. We changed how route parsing is performed in 0.15 and I remember encountering this issue myself.

I might have already pushed a fix to master or another branch. Anyway, I will take a look at it very soon (hopefully tomorrow). :-)

**Comment by florimondmanca** [comment]
@DariusMiu Okay so, I looked this up. You should be able achieve what you want using `req.url.path`:

```python
import bocadillo
api = bocadillo.App()

@api.route('{}')
async def slash(req, res) :
    print("do stuff with", req.url.path)
```

See also [Request URL](https://bocadilloproject.github.io/guides/http/requests.html#url).

Considering this, I don't think this requires modifications to the code, but I think we can add this trick to the docs.

**Comment by DariusMiu** [comment]
@florimondmanca Yeah, I figured that out, too. that requires additional parsing, though. in my example, I have file handlers for things ending in .png and the like, and didn't want to check the url a second time

**Comment by florimondmanca** [comment]
I'm not sure I understand your use case. Can you share the exact route pattern you're using? I suspect it would be something like `/images/{path_to_image}.png`, where `path_to_image` should be able to contain slashes, right?

**Comment by DariusMiu** [comment]
@florimondmanca pretty much that. here's my code: https://gitlab.com/kheina/kheina.com/blob/master/server.py#L112

**Comment by florimondmanca** [comment]
Thanks for the link.

I see you're retrieving the path in the URL in order to read the corresponding file from disk.

Have you considered using [static files](https://bocadilloproject.github.io/guides/http/static-files.html) instead? If so, what prevented you from using it?

**Comment by florimondmanca** [comment]
Hi @DariusMiu, is this still an issue for you, or have you found a workaround?

**Comment by DariusMiu** [comment]
Yes, but I haven't been working on it actively. Have a couple other projects plus just started a new job, so quite busy. I'll look into it this weekend and report back!

**Comment by DariusMiu** [comment]
I can get things working with static files well enough, though I can't set how long to cache things for specific file types, or set headers for file types. It also seems as though errors raised through static files aren't handled by the error handler specified for `bocadillo.HTTPError`

**Comment by DariusMiu** [comment]
Additionally, if a static file is changed while the server is still running, I get the following error:
```
ERROR: Exception in ASGI application
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/bocadillo/applications.py", line 242, in dispatch
    return await app(scope, receive, send)
TypeError: __call__() takes 3 positional arguments but 4 were given

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/uvicorn/protocols/http/httptools_impl.py", line 368, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/usr/lib/python3.6/site-packages/bocadillo/applications.py", line 257, in __call__
    await self.asgi(scope, receive, send)
  File "/usr/lib/python3.6/site-packages/starlette/middleware/trustedhost.py", line 51, in __call__
    await self.app(scope, receive, send)
  File "/usr/lib/python3.6/site-packages/bocadillo/applications.py", line 245, in dispatch
    return await app(scope, receive, send)
  File "/usr/lib/python3.6/site-packages/starlette/middleware/wsgi.py", line 62, in __call__
    await responder(receive, send)
  File "/usr/lib/python3.6/site-packages/starlette/middleware/wsgi.py", line 90, in __call__
    await asyncio.wait_for(sender, None)
  File "/usr/lib64/python3.6/asyncio/tasks.py", line 339, in wait_for
    return (yield from fut)
  File "/usr/lib/python3.6/site-packages/starlette/middleware/wsgi.py", line 105, in sender
    await send(message)
  File "/usr/lib/python3.6/site-packages/uvicorn/protocols/http/httptools_impl.py", line 489, in send
    raise RuntimeError("Response content longer than Content-Length")
RuntimeError: Response content longer than Content-Length
```

**Comment by florimondmanca** [comment]
Thanks for reporting back!

If the static plugin/helper is not enough, you can try building WSGI WhiteNoise apps directly (see [here](http://whitenoise.evans.io/en/stable/base.html)), one for each file type, and then [mount](https://bocadilloproject.github.io/guide/nested-apps.html) them as a nested WSGI apps. (The docs say that nested apps should be either Bocadillo or ASGI apps, but WSGI is actually also supported.)

As for the http_tools runtime error — that’s interesting... I’ll have to try to reproduce this.

**Comment by DariusMiu** [comment]
code is up on [gitlab](https://gitlab.com/kheina/kheina.com/blob/master/server.py#L58), you can see the error occurring here:
https://kheina.com/page/that/does/not/exist
https://kheina.com/static/file/that/does/not/exist.png

Thanks for the WSGI tip, I'll look into it!

**Comment by florimondmanca** [comment]
@DariusMiu I was able to reproduce the `RuntimeError` issue :+1:  see #288.

**Comment by florimondmanca** [comment]
@DariusMiu I will be releasing a fix in 0.16.1 (see #299) which will allow you to have `/{path:path}.png` (note the `:path` converter) match `/images/image.png`. This is potentially impacting you so I thought I'd raise notice.

**Comment by florimondmanca** [comment]
@DariusMiu I am closing this for housekeeping purposes, since I believe the fix mentioned in my previous comment should allow you to achieve what you initially asked. But feel free to comment again if you've been lucky with manually adding WSGI apps. :-)
